### ISSUE TITLE
Introduce an index for thing type definitions

### ISSUE THREAD

**Issue by kaikreuzer** [eclipse/smarthome#1729]
Analyzing performance issues on the Paper UI lately, I noticed that the REST API response on /thing-types is about 1.5MB with Z-Wave binding installed and 62KB without.

This clearly indicates that there are some bindings with a huge number of thing types and that we need to optimize the way to access this information.

A while ago, @cdjackson already predicted this and suggested a kind of index of thing types within the bindings. This index should only have the most relevant information (like id, label, description, what else?) and we should have a REST url to access only this information. If the full glory is required, this should be specifically requested for a (or a list of) given thing type UID(s).

Ideally, we would even generate this index during the build process, but besides optimizing the UI access, it will also reduce the startup time as the full XMLs should only be parsed upon request and not upfront anymore.

**Comment by cdjackson** [comment]
> (like id, label, description, what else?) 

For ZWave, I need a few bits of device specific information that I use to compare with the information recovered from the device. I suspect that the information required will be binding dependant, so either we need a generic way to describe this, or it needs to be flexible. Providing a generic solution might be difficult as it ultimately results in having to modify things to fit the generic description, and it’s therefore not optimum.

The information in ZWave that I need here is stored in the thing properties - there are 3 properties from memory - manufacturer information, device information and version.

I like the idea of auto generating this - do you have an idea how? Maybe we can add an attribute in the XML to say what properties should be added to the index - this makes it flexible for different bindings.

**Comment by kaikreuzer** [comment]
I assume you need this information for the discovery service to decide for the right type, but not have these things exposed to the UI through the REST API, right?

In case the content is binding specific (do you have other examples than ZWave, which will need it?), we will also have to think about how this information is provided - will key-value pairs be enough?

Besides generation at build time (which can often be a hassle as we see with the DSL generation steps), another option might also be to create the index on the first start of a bundle - we could even use the storage service (i.e. the database) to store the index in it. Might be a valid option as well.

**Comment by cdjackson** [comment]
> I assume you need this information for the discovery service to decide for the right type,
> 
Correct.

> but not have these things exposed to the UI through the REST API, right?
> 
I'm not sure what you mean by this? If you mean do I need the index exposed, then that’s correct - the index doesn’t need to be exposed through the REST API.

> In case the content is binding specific (do you have other examples than ZWave, which will need it?)
> 
Zigbee might, although at the moment it’s written avoid this since the ZCL is more self describing than ZWave. So, no, no other good examples at the moment.


> , we will also have to think about how this information is provided - will key-value pairs be enough?
> 

In ZWave, the data is stored in the properties -:
      <property name="manufacturerId">0001</property>
      <property name="manufacturerRef">7FFF:7FFF</property>

In some instances there is also version numbers. So, key-value pairs is fine - same as the properties currently is.


> Besides generation at build time (which can often be a hassle as we see with the DSL generation steps), another option might also be to create the index on the first start of a bundle
> 
Yes - this could be a good option.

My suggestion for auto generation might look like this -:
      <property index name="manufacturerId">0001</property>
      <property index name="manufacturerRef">7FFF:7FFF</property>

i.e. just adding the ‘index’ attribute. This could allow the index to be generated at runtime or compile - to the user it could be the same, but I agree runtime might be better.

> - we could even use the storage service (i.e. the database) to store the index in it. Might be a valid option as well.
> 
If by this you mean to persist the data then I’m not sure this is a good idea unless you can determine changes. In ZWave at least, these IDs do change regularly as manufacturers add new versions that have the same properties (normally it’s used for regional differences). So, you might just be better to create the index every time and store in memory.

Of course, creating at runtime doesn’t solve the startup time issue - it might reduce memory usage since in the end you only need to retain the index in memory, but the startup will be slower than now I guess.

**Comment by kaikreuzer** [comment]
> This index should only have the most relevant information (like id, label, description, what else?) and we should have a REST url to access only this information. If the full glory is required, this should be specifically requested for a (or a list of) given thing type UID(s).

Actually, I think we can also split the implementation in two separate steps here:

1. Adapt the REST API to only return the list of thing types with reduced information and let the UIs request the "full glory" through an individual REST call for a specific thing type. This step should be easy and it would speed up the Paper UI right away (since I have heard many complaints about loading speed due to the call to read all thing-types).
2. Improve the loading of the thing type XMLs by introducing an index, which is used to answer the request for the list of all thing types and where the XML is only loaded if an individual thing type is requested.

What we would need to define for step 1 is the right content, though - this should already keep in mind what information we could/want provide in the index later on.
@cdjackson & @aounhaider1 What information would be required by the UI for the "full list"? Would it be the current response minus the config description and channels or can anything further be left out?

**Comment by Mixajlo** [comment]
I am about to start working on this issue. So, as @kaikreuzer suggested, I will remove the configDescription and the channels for now, but if you guys  ( @cdjackson , @aounhaider1 ) have something to add/remove, please let me know.

**Comment by aounhaider1** [comment]
One thing, removing channels from thingType will cause the control page of PaperUI to send a lot more request to server. Would that be fine?

**Comment by kaikreuzer** [comment]
Yes - one request per thing-type on the page, right?
In case we find this too inefficient, we could also provide the option to request a list of thing types within a single request, but I would first do without it.

**Comment by aounhaider1** [comment]
Yes, for a few things it will be fine. But it will surely affect the performance if there are too many things.
List based request would be a good solution in my opinion.

**Comment by cdjackson** [comment]
Just to clarify my understanding - sorry for any repetition but I’m interested in making sure this meets the needs of ZWave since it’s the biggest ‘offender’ with the number of XML files ;)…

The index is there so that the system doesn’t need to load all the thing XML files on startup. It will provide a list of things in a single file, and will allow bindings and UIs to load this list (right?). Once a binding identifies that it needs a specific thing definition (presumably when the thing is instantiated, or requested via REST) then the full thing XML definition will be loaded….

The REST interface will allow a list of things (e.g. thing type, label, binding), but no detail until the thing type XML is loaded (as per Kais list above).

Is this correct?

If so, then for me parameters and channels aren’t needed in the list of things. I don’t believe the UI needs this information until after then thing is instantiated - for sure HABmin doesn’t - it only loads the thing type for a thing that is being edited so there’s no need I think for a list with multiple thing types.

As mentioned earlier I need the properties (or at least a subset of properties) to be present in the index.

**Comment by kaikreuzer** [comment]
> Is this correct?

Yes, more or less. What you have to keep in mind is that XML loading and REST responses are two different issues (that's why I suggested to split this issue into two): Even without any XML files, a binding could dynamically create 1000 thing types and thus the REST response would be too huge. So it is not "no detail until the thing type XML is loaded", but "no detail until the thing type is specifically requested by the UI".

> As mentioned earlier I need the properties (or at least a subset of properties) to be present in the index.

In the index, right, in order to figure out, which one you want to load. But do you also need it in the REST response for all thing types?

**Comment by cdjackson** [comment]
> So it is not "no detail until the thing type XML is loaded", but "no detail until the thing type is specifically requested by the UI".

Understood.

> In the index, right, in order to figure out, which one you want to load. 

Correct.

> But do you also need it in the REST response for all thing types?

Probably it's not needed in the REST response - I can't think of a use case at the moment at least.

**Comment by cdjackson** [comment]
Is this meant to be closed. It looks like it was closed accidentally with a 'fixes part of' statement....
